### DISCUSSION TITLE
MultiPV implementation

### DISCUSSION THREAD

**Discussion by dav1312** [official-stockfish/Stockfish#5230] [Ideas]
Following the conversation that happened on Discord about MultiPV I decided to open a formal discussion here on GitHub.
There are currently 3 ways to implement MultiPV:
1. Equal amount of resources for all moves (current)
	- Useful if you really want to analyze all moves equally, but it drastically reduces the quality of the evaluation by wasting valuable resources on bad moves.
2. Return the value for each root move (Fake/free MultiPV)
	- This doesn't cost nearly as many resources as the other two methods. It doesn't change the behavior of the engine compared to using just 1 line, and it can provide some insight into what the engine is "thinking" about the moves when performing a regular search. I suspect this is how many (non-power) users already assume that MultiPV works. Not changing the engine's behavior can be a problem if the user actually wants to force the engine to search for other moves.
3. Adjust amount of resources for each line (https://github.com/official-stockfish/Stockfish/pull/2152)
	- This is a middle ground between the other two implementations. Adjusting the resources based on various factors such as line number, evaluation, or winning chances compensates for the issue of wasting resources. There are _a lot_ of users who use the engine in the browser in popular websites that I'm sure are unaware of the strength penalty that comes with using a lot of lines or, if they are aware, don't reduce them because they want to see more options than just the best move. This method ensures that the user doesn't need to change anything, the engine decides for them how to spend the resources.

**Comment by vdbergh** [comment]
2. is bad I think. Evals outside the window are totally unreliable.

**Comment by jhellis3** [comment]
I've been testing a few things locally.  What I came up with was -2 depth at root for pvIdx > 0 and allowing TT on PvNodes when pvIdx > 0 and it is our move with a lower bound. 

The combination of these two things means that we are able to use the previous iteration's main line results to very efficiently skip work on trivial transpositions.  These transpositions currently waste a great deal of resources because they are going to have the same score as the main line, and are first in line for multiPV while contributing nothing useful.  Taking the TT cutoff for our own moves with a lower bound is safe because it is either a fail high, which if large enough will promote to mainline and be resolved as ususal, or it has BOUND_EXACT which means it was searched as a PvNode at a greater depth than our current search.

At the same time, it preserves the spirit of multiPV in that all of our opponents moves are disallowed from taking nonPV cutoffs/reductions. 

I think this was worth ~48 Elo at an STC of 18+0.18, MPV=4 for both.  I can push a branch and test on fishtest if this is something people are interested in...

**Reply by jhellis3** [reply]
Here is the patch: https://github.com/jhellis3/Stockfish/commit/2236edbc626ceba35e3d7995923cb76b4a08873c

**Comment by mstembera** [comment]
In SMP different threads often find different moves.   get_best_thread() is the function that then picks one of them.  We could output all of them in MultiPV w/o spending any additional resources.  As a bonus this happens most often in positions where multiple moves are close which is when most people care about MultiPV.

**Comment by bftjoe** [comment]
Another idea:

If number of threads >= multipv, have each thread search different moves at root. could partly use existing searchmoves infrastructure to handle this, with possible optimization for searchmoves of only 1 move. Leftover threads could just search best move or be equally distributed between all multipv moves.

**Comment by dubslow** [comment]
Seems that part of the problem is that we're not sure what users want, which is a key piece of data to make a good choice. I can think of at least two broadly different user bases, the casuals who use browsers (e.g. lichess or chesscom), and the "power" users of correspondence. Soliciting from feedback from at least the latter if not the former would be useful.

It would also be good to perhaps explore the options in more detail. The "overkill" way would be to implement *all* the choices, make it selectable via UCI and thence solicit feedback from power users or browser users. However, maintainers have a history of disdaining option overload, so mayhaps this would need to be a branch outside (but tracking) master.

That said, for browser users at least, using a fake/free/pseudo mode is likely to be a clear improvement.

Also, if I interpret jhellis3 correctly (not sure I do), that sounds like a straight, direct improvement to the current "equal time" strategy, and so quite separately from the question of what other modes are good, improving the current mode is certainly worth a fishtest.

**Comment by jhellis3** [comment]
Fishtest result for aforementioned patch can be found here: https://tests.stockfishchess.org/tests/view/663e5d163a2f9702074bc16a

**Comment by PGG106** [comment]
Just as i said on Discord, i'm not sure STC gains from an option that should be turned off in regular game playing anyway make for a good argument for any type of tweak.
MultiPV is an "analysis" tool that comes with a clear and understood impact on strength (given the same amount of time per search at least), users should be aware of this and if they aren't it's not really an issue with the feature itself.
Any form of "multiPV reduction" will bridge the gap in strength, sure, but how do we measure if the tweak doesn't defeat the point of multipv? 
You can reduce the depth of the secondary lines all the way down to 0 and gain Elo, and so what? How do we establish that -2 is ok but -20 isn't? how are we sure that "-2" in itself is acceptable when there's no clear way to measure "analysis" performance, with analysis being the only realistic usecase for multipv? 
tldr: If an end user doesn't want to lose Elo to multipv they should just turn it off, Elo gains tweaks are  meaningless for a feature that shouldn't actively care about Elo at all.

**Reply by bftjoe** [reply]
I disagree that it is a clear and understood impact when chess.com defaults to multipv 3 for analysis. I highly doubt that most people using this tool know what impact this has on analysis strength.

**Reply by PGG106** [reply]
That points to a need to better educate the end user, "acting behind Their back" by changing the default behaviour of multipv (to in practice do less multipv) is a non solution imo, I just don't think the default behaviour of a function like multipv should try to go out of their way to account for misuage by ignorant end users

**Reply by Craftyawesome** [reply]
Default is comparatively unimportant compared to what and how to add an alternative IMO, but worth discussing at least.
Whether the default is traditional multipv or not, whatever the option is named could help quite a bit. Something like "Fast MultiPV" or "Smart MultiPV" would both hint power users that something could have changed and hint novices that something about multipv is worth an option. 
Also not sure I would call it "less multipv". AFAIK searching every line to equal depth isn't inherent to the definition of multipv, just that equal depth is how it is traditionally and usually done.

**Comment by Craftyawesome** [comment]
You could also argue for a variant of 1 with something like a node limit, since sometimes other moves take more nodes to resolve the same depth. I suspect this is actually a decent portion of multipv elo loss.

IMO 2 would be nice to have if you can keep it as decent as leela's output. It also makes the most sense, since it considers the secondary moves the right amount for maximum elo and naturally looks at bad moves less. But I haven't seen this done well in an AB, and there are concerns like out of window searches mentioned above.

3 could be trying to simulate 2 if 2 can't work directly (maybe search at the adjusted depth but with new window? Would be some level of elo loss though)

Personally I don't see much of an appeal between "minimal elo loss with usable mpv" and "full search", but if someone does want it you may want to use some UCI option to control the balance between it.

**Comment by bftjoe** [comment]
One thing to consider is that stockfish/uci support setting a maximum thinking time/nodes/depth/etc but not a minimum. Support for a minimum would be interesting/useful.

**Comment by joergoster** [comment]
Maybe this is of some interest: https://github.com/joergoster/Stockfish/releases/tag/mpv1
The corresponding branch https://github.com/joergoster/Stockfish/tree/simpleMultiPV

**Comment by bftjoe** [comment]
I have a simple idea: only main thread searches multiple PV moves. This should dramatically reduce the strength loss of enabling MultiPV if multiple threads are used. Let me know what you think.

https://github.com/bftjoe/Stockfish/tree/multipv

**Comment by xu-shawn** [comment]
These are all great ideas. But even if we do not take into account the aspect the distributing search effort on the best move, I still consider the current MultiPV implementation inefficient. Specifically, there are two ways where effort is completely wasted:

* On each root search, a significant portion of effort is spent refuting alternative variations. However, since each MultiPV iteration are searching on the same root position, a lot of moves are refuted many times across a MultiPV depth. This is a significant duplication of effort.
* On each root search, we can end up with more than one PV and exact score (through raising alpha multiple times). However, only the score on the last alpha raise is kept. The rest are discarded.

I believe we can eliminate these inefficiencies with a single-iteration MultiPV just through modifying the root search and aspiration windows. To illustrate the difference, this is a normal implementation of aspiration windows:

1. start with window `[a, b]` and `delta`
2. Perform a root search with `[a, b]` window, let `v` be the score returned by the search
3. if `v <= a`, set `a := v - delta`. Increase `delta`. Go to step 2
4. if `v >= b`, set `b := v + delta`. Increase `delta`. Go to step 2
5. Otherwise, `a < v < b`, the score is exact. `v` corresponds to a valid PV. Report and go the next depth

We can modify aspiration windows for MultiPV=n. The modified procedure would be as follows:

1. start with window `[a, b]` and `delta`
2. Perform a root search with `[a, b]` window. Do not increase alpha at root. Let `v0, v1, ..., vn` be the `n` highest score found at root, at descending order.
3. if `vn <= a`, set `a := vn - delta`. Increase `delta`. Go to step 2
4. if `v0 >= b`, set `b := v0 + delta`. Increase `delta`. Go to step 2
5. Otherwise, `a < vn <= ... <= v0 < b`. All `v0` to `vn` scores correspond to a valid PV. Report all scores and go to the next depth

Ideally, this would not result in as large of a drop in strength, especially when the top moves have similar scores. Additionally, since all PV nodes are searched at similar depth, this probably would not reduce the quality of alternative moves.

**Reply by xu-shawn** [reply]
Actually it's possible to optimize this even further through raising alpha to the n-th highest root score, after n such valid scores have been found.

**Reply by cj5716** [reply]
This is something I had in mind, and one that I believe should be implemented into master.

**Comment by joergoster** [comment]
This sounds more like a possible refinement (or not) of the apiration window(s) in case of a MultiPV search.

Imho, we should first discuss how the basic implementation of a MultiPV search in a _modern_ AB engine with PVS (Principal Variation Search) should look like.

Here is my take:

1. All PV lines should be searched one after another and in the order of the last iteration's result. And only after the last PV line, the remaining Non-PV moves should be searched with a zero window. To me it makes no sense to search a root move first as a Non-PV move, and this possibly many times depending on the number of requested PV lines, only to later search it as a PV move anyways.
 
2. Inferior lines don't must be searched to the same depth as the best line(s). This looks like wasted effort to me, especially in the light of how MCTS engines work. And I don't recall a single complaint about the MultiPV mode of, let's say, Lc0. (Reason for this might be that almost nobody is using Lc0 in MultiPV mode, idk ...)

**Comment by dav1312** [comment]
Final results of the user survey regarding MultiPV
![image](https://github.com/user-attachments/assets/ad406ee5-cc00-48b1-8f43-826fdac7296f)
![image](https://github.com/user-attachments/assets/1551809d-1c61-494b-9513-8e442ec7d31c)
![image](https://github.com/user-attachments/assets/92739501-f616-49bc-bf6e-d9ebe15a8d99)

**Reply by cj5716** [reply]
I think the last question could be a bit ambiguous, maybe a better phrasing would be "would you prefer Stockfish to distribute resources equally between the N lines dictated by MultiPV, or would you prefer that the worse lines of the N had fewer resources assigned to them?"

**Comment by vondele** [comment]
I think the survey suggests that indeed our current scheme is not really aligned with user expectations, and that we can do better. What is missing is a precise and measurable target to develop the new scheme with. I believe we need to aspects:

* A target on the cost of multiPV
* A benchmark for the quality of multiPV.

To clarify these by examples.

* using multiPV 5 should not reduce the Elo strength in a match against master without multiPV by more than 50 Elo (e.g. STC 10k games)
* on a testset of 1000 positions picked randomly from fishtest games, we pick the method that scores most points, and we give points by looking at the found moves in a 1sec / pos multiPV 5 search giving 5,4,3,2,1 points if the 1,2,3,4,5th move matches a high level reference multiPV search.

There might be additional aspects (e.g. if the scheme generalizes naturally to multiPV N, is simple, sound, etc). After such a quantitative benchmark is established, we can benchmark available PRs against it.
